bytecomp/bytelink.ml

```diff
@@ -227,46 +228,44 @@ let link_compunit output_fun currpos_fun inchan file_name compunit =
     debug_info := (currpos_fun(), debug_event_list, debug_dirs) :: !debug_info
   end;
   output_fun code_block;
-  if !Clflags.link_everything then
-    List.iter Symtable.require_primitive compunit.cu_primitives
+  let fold_primitive needs_stdlib name =
+    if !Clflags.link_everything then
+      Symtable.require_primitive name;
+    (needs_stdlib || name = "%standard_library_default")
+  in
+  List.fold_left fold_primitive accu compunit.cu_primitives
 
 (* Link in a .cmo file *)
 
-let link_object output_fun currpos_fun file_name compunit =
-  let inchan = open_in_bin file_name in
-  try
-    link_compunit output_fun currpos_fun inchan file_name compunit;
-    close_in inchan
-  with
-    Symtable.Error msg ->
-      close_in inchan; raise(Error(Symbol_error(file_name, msg)))
-  | x ->
-      close_in inchan; raise x
+let link_object accu output_fun currpos_fun file_name compunit =
+  In_channel.with_open_bin file_name @@ fun inchan ->
+    try link_compunit accu output_fun currpos_fun inchan file_name compunit
+    with Symtable.Error msg -> raise(Error(Symbol_error(file_name, msg)))
 
 (* Link in a .cma file *)
 
-let link_archive output_fun currpos_fun file_name units_required =
-  let inchan = open_in_bin file_name in
-  try
-    List.iter
-      (fun cu ->
+let link_archive accu output_fun currpos_fun file_name units_required =
+  In_channel.with_open_bin file_name @@ fun inchan ->
+    List.fold_left
+      (fun accu cu ->
          let n = Compunit.name cu.cu_name in
          let name = file_name ^ "(" ^ n ^ ")" in
          try
-           link_compunit output_fun currpos_fun inchan name cu
+           link_compunit accu output_fun currpos_fun inchan name cu
          with Symtable.Error msg ->
            raise(Error(Symbol_error(name, msg))))
-      units_required;
-    close_in inchan
-  with x -> close_in inchan; raise x
+      accu units_required
 
 (* Link in a .cmo or .cma file *)
 
-let link_file output_fun currpos_fun = function
+let link_file output_fun currpos_fun accu = function
     Link_object(file_name, unit) ->
-      link_object output_fun currpos_fun file_name unit
+      link_object accu output_fun currpos_fun file_name unit
   | Link_archive(file_name, units) ->
-      link_archive output_fun currpos_fun file_name units
+      link_archive accu output_fun currpos_fun file_name units
+
+let link_files output_fun currpos_fun =
+  List.fold_left (link_file output_fun currpos_fun) false
```

Comment:
```suggestion
  List.exists (link_file output_fun currpos_fun)
```

Reply:
That's not equivalent, though - we do need to link all of the files, regardless of whether they use that symbol or not!